### PR TITLE
feat: added groq provider

### PR DESCRIPTION
**Pull request by lifeizhou-ap**
- Added groq provider
- Converted content object list in message (ToolResponse message) to string to match api schema only for Ollama and Groq. 
- Removed the duplications in providers that compatible wit openai api schema
- Extracted open ai related utils function to a new file `openai_utils`
- Added integration tests for google and groq providers and extract the utils to mock_server.rs

Note:
I could not get the capture screenshot working with groq provider as when the message include the toolResult with image the api returns bad request. So far I found they can categorised into 2 scenarios as below:
-  For most non-vision models, it does not accept the user message with image although in the [doc](https://console.groq.com/docs/api-reference#chat) it says it does.  

   Error message:
`{
  "request_id": "req_01jfckkap9fcdv2c4cr4tepnfw",
  "created_at": "2024-12-18T10:04:15.434Z",
  "error": {
    "message": "message[0].content must be a string",
    "type": "invalid_request_error"
  }
}`

- I also tried `llama-3.2-11b-vision-preview`, it does not accept the payload including the user message with image and system message at the same time.
   Error message:
`{
  "request_id": "req_01jfcmfx6yf0zbjt4028v4gkcd",
  "created_at": "2024-12-18T10:19:51.904Z",
  "error": {
    "message": "prompting with images is incompatible with system messages",
    "type": "invalid_request_error"
  }
}`

I tried with the python cli, it does not work either but with different reason. I did not look in-depth.  I feel this feature does not work originally and we can enhance this in the future when we have more usages with groq

### CODE REVIEW

File: crates/goose/src/providers/openai_utils.rs [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+                                            "role": "user",
+                                            "content": [convert_image(&image, image_format)]
+                                        }));
+                                    }
+                                    _ => {
+                                        tool_content.push(content);
+                                    }
+                                }
+                            }
+                            let concatenated_content = tool_content
```
Comment (lifeizhou-ap) [inline crates/goose/src/providers/openai_utils.rs]: Converted content object list in message (ToolResponse message) to string to match api schema.   In [openai](https://platform.openai.com/docs/api-reference/chat/create), it accepted "string" or "array", but in groq it only accepts "string" based on their [api](https://console.groq.com/docs/api-reference#chat-create)

File: crates/goose/src/providers/factory.rs [hunk 2]
```diff
@@ -3,6 +3,7 @@ use super::{
     databricks::DatabricksProvider, google::GoogleProvider, ollama::OllamaProvider,
     openai::OpenAiProvider,
 };
+use crate::providers::groq::GroqProvider;
```
Comment (zakiali) [inline crates/goose/src/providers/factory.rs]: prob move this up into the super

File: crates/goose/src/providers/databricks.rs [hunk 3]
```diff
@@ -8,11 +8,12 @@ use super::base::{Provider, ProviderUsage, Usage};
...
-use super::utils::{
-    check_bedrock_context_length_error, check_openai_context_length_error, get_model,
-    messages_to_openai_spec, openai_response_to_message, tools_to_openai_spec,
-};
+use super::utils::{check_bedrock_context_length_error, get_model};
 use crate::message::Message;
+use crate::providers::openai_utils::{
+    check_openai_context_length_error, messages_to_openai_spec, openai_response_to_message,
+    tools_to_openai_spec,
+};
```
Comment (zakiali) [inline crates/goose/src/providers/databricks.rs:16]: nit: we should prob use the get_openai_usage and handle_response in this one as well.

File: crates/goose-cli/src/commands/configure.rs [hunk 4]
```diff
@@ -158,7 +161,8 @@ pub fn get_recommended_model(provider_name: &str) -> &str {
         "databricks" => "claude-3-5-sonnet-2",
         "ollama" => OLLAMA_MODEL,
         "anthropic" => "claude-3-5-sonnet-2",
-        "google" => "gemini-1.5-flash",
+        "google" => GOOGLE_DEFAULT_MODEL,
```
Comment (zakiali) [inline crates/goose-cli/src/commands/configure.rs:167]: nit: this seems like something we should standardize in the providers for openai, databricks, and anthropic

File: crates/goose/src/providers/openai_utils.rs (hunk 1 above)
Comment (zakiali) [reply crates/goose/src/providers/openai_utils.rs]: @baxen is this related to the error you were seeing on the llama3 databricks endpoint?

File: crates/goose/src/providers/groq.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+impl GroqProvider {
+    pub fn new(config: GroqProviderConfig) -> anyhow::Result<Self> {
+        let client = Client::builder()
+            .timeout(Duration::from_secs(600)) // 10 minutes timeout
+            .build()?;
+
+        Ok(Self { client, config })
+    }
+
+    fn get_usage(data: &Value) -> anyhow::Result<Usage> {
```
Comment (zakiali) [inline crates/goose/src/providers/groq.rs:31]: we should consider adding get_usage as a trait for Providers in the base object

File: crates/goose-cli/src/commands/configure.rs (hunk 4 above)
Comment (lifeizhou-ap) [reply crates/goose-cli/src/commands/configure.rs:167]: done

File: crates/goose/src/providers/databricks.rs (hunk 3 above)
Comment (lifeizhou-ap) [reply crates/goose/src/providers/databricks.rs:16]: done

File: crates/goose/src/providers/groq.rs (hunk 5 above)
Comment (lifeizhou-ap) [reply crates/goose/src/providers/groq.rs:31]: agree. This PR has many changes. I can create a follow-up PR after this PR is merged

**Comment by michaelneale** [comment]
@lifeizhou-ap I wouldn't worry about image support for groq and llama models for now if they are proving a problem.

File: crates/goose/src/providers/openai_utils.rs (hunk 1 above)
Comment (lifeizhou-ap) [reply crates/goose/src/providers/openai_utils.rs]: I've changed to only apply the concatenate logic to ollama and groq

File: crates/goose-server/src/configuration.rs [hunk 6]
```diff
@@ -88,6 +89,17 @@ pub enum ProviderSettings {
         #[serde(default)]
         max_tokens: Option<i32>,
     },
+    Groq {
```
Comment (zakiali) [inline crates/goose-server/src/configuration.rs:92]: one thing I missed in the google provider and here initially was that we also want to add context_limit and estimate_factor here as well

File: crates/goose-server/src/configuration.rs [hunk 7]
```diff
@@ -168,6 +181,19 @@ impl ProviderSettings {
...
+                host,
+                api_key,
+                model,
+                temperature,
+                max_tokens,
+            } => ProviderConfig::Groq(GroqProviderConfig {
+                host,
+                api_key,
+                model: ModelConfig::new(model)
+                    .with_temperature(temperature)
```
Comment (zakiali) [inline crates/goose-server/src/configuration.rs:194]: context_limit and estimate_factor should go here as well for Groq and Google

File: crates/goose/src/providers/groq.rs [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    }
+
+    async fn complete(
+        &self,
+        system: &str,
+        messages: &[Message],
+        tools: &[Tool],
+    ) -> anyhow::Result<(Message, ProviderUsage)> {
+        let payload =
+            create_openai_request_payload(&self.config.model, system, messages, tools, true)?;
```
Comment (zakiali) [inline crates/goose/src/providers/groq.rs:65]: the boolean argument here (true) is a little confusing. I don't think there is a way to specify kwargs, but maybe setting a var before calling the function (e.g. concat_tool_response_contents=true) and passing that in to the function call would be a little more clear?

**Review by zakiali — APPROVED** [review]
looks good! left a few comments

**Pull request merged**